common/map.h

```diff
@@ -0,0 +1,462 @@
+  using ImplT =
+      RawHashtable::TableImpl<MapBase<InputKeyT, InputValueT>, SmallSize>;
+
+ public:
+  using KeyT = typename BaseT::KeyT;
+  using ValueT = typename BaseT::ValueT;
+
+  Map() = default;
+  Map(const Map& arg) = default;
+  template <ssize_t OtherMinSmallSize>
+  explicit Map(const Map<KeyT, ValueT, OtherMinSmallSize>& arg) : ImplT(arg) {}
+  Map(Map&& arg) = default;
+  template <ssize_t OtherMinSmallSize>
+  explicit Map(Map<KeyT, ValueT, OtherMinSmallSize>&& arg)
+      : ImplT(std::move(arg)) {}
+
+  // Reset the entire state of the hashtable to as it was when constructed,
+  // throwing away any intervening allocations.
+  void Reset();
+};
+
+template <typename InputKeyT, typename InputValueT>
+template <typename LookupKeyT>
+auto MapView<InputKeyT, InputValueT>::Contains(LookupKeyT lookup_key) const
+    -> bool {
+  return this->LookupEntry(lookup_key) != nullptr;
+}
+
+template <typename KT, typename VT>
+template <typename LookupKeyT>
+auto MapView<KT, VT>::Lookup(LookupKeyT lookup_key) const -> LookupKVResult {
+  return LookupKVResult(this->LookupEntry(lookup_key));
+}
+
+template <typename KT, typename VT>
+template <typename LookupKeyT>
+auto MapView<KT, VT>::operator[](LookupKeyT lookup_key) const -> ValueT* {
+  auto result = Lookup(lookup_key);
+  return result ? &result.value() : nullptr;
+}
+
+template <typename KT, typename VT>
+template <typename CallbackT>
+void MapView<KT, VT>::ForEach(CallbackT callback) {
+  this->ForEachEntry(
+      [callback](EntryT& entry) { callback(entry.key(), entry.value()); },
+      [](auto...) {});
+}
+
+template <typename KT, typename VT>
+template <typename LookupKeyT>
+[[clang::always_inline]] auto MapBase<KT, VT>::Insert(LookupKeyT lookup_key,
+                                                      ValueT new_v)
+    -> InsertKVResult {
+  return Insert(lookup_key,
+                [&new_v](LookupKeyT lookup_key, void* key_storage,
+                         void* value_storage) -> std::pair<KeyT*, ValueT*> {
+                  KeyT* k = new (key_storage) KeyT(lookup_key);
+                  auto* v = new (value_storage) ValueT(std::move(new_v));
+                  return {k, v};
+                });
+}
+
+template <typename KT, typename VT>
+template <typename LookupKeyT, typename ValueCallbackT>
+[[clang::always_inline]] auto MapBase<KT, VT>::Insert(LookupKeyT lookup_key,
+                                                      ValueCallbackT value_cb)
+    -> InsertKVResult
+  requires(
+      !std::same_as<ValueT, ValueCallbackT> &&
+      std::convertible_to<decltype(std::declval<ValueCallbackT>()()), ValueT>)
+{
+  return Insert(lookup_key,
+                [&value_cb](LookupKeyT lookup_key, void* key_storage,
+                            void* value_storage) -> std::pair<KeyT*, ValueT*> {
+                  KeyT* k = new (key_storage) KeyT(lookup_key);
+                  auto* v = new (value_storage) ValueT(value_cb());
```

Comment:
Why use `auto` here but not the previous line?
```suggestion
                  ValueT* v = new (value_storage) ValueT(value_cb());
```